### PR TITLE
Introduce separate shard limit for frozen shards

### PR DESCRIPTION
**Pull request by henningandersen**
Frozen indices (partial searchable snapshots) require less heap per
shard and the limit can therefore be raised for those. We pick 3000
frozen shards per frozen data node, since we think 2000 is reasonable
to use in production.

Relates #71042 and #34021

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/indices/ShardLimitValidator.java [hunk 1]
```diff
@@ -81,55 +111,119 @@ public void validateShardLimit(final Settings settings, final ClusterState state
...
      * @param state             The current cluster state
      * @return If present, an error message to be given as the reason for failing
      * an operation. If empty, a sign that the operation is valid.
      */
-    public Optional<String> checkShardLimit(int newShards, ClusterState state) {
-        return checkShardLimit(newShards, state, getShardLimitPerNode());
+    private Optional<String> checkShardLimit(int newShards, int newFrozenShards, ClusterState state) {
+        // we verify the two limits independently. This also means that if they have mixed frozen and other data-roles nodes, such a mixed
+        // node can have both 1000 normal and 3000 frozen shards. This is the trade-off to keep the simplicity of the counts. We advocate
+        // against such mixed nodes for production use anyway.
```
Comment (jasontedor) [inline server/src/main/java/org/elasticsearch/indices/ShardLimitValidator.java:178]: I agree with this, good comment.

File: server/src/main/java/org/elasticsearch/indices/ShardLimitValidator.java [hunk 2]
```diff
@@ -81,55 +111,119 @@ public void validateShardLimit(final Settings settings, final ClusterState state
...
+
+    private static int nodeCount(ClusterState state, Predicate<DiscoveryNode> nodePredicate) {
+        return (int)
+            StreamSupport.stream(state.getNodes().getDataNodes().values().spliterator(), false)
+                .map(oc -> oc.value).filter(nodePredicate).count();
+    }
+
+    private static boolean hasFrozen(DiscoveryNode node) {
+        final List<DiscoveryNodeRole> dataRoles =
+            node.getRoles().stream().filter(DiscoveryNodeRole::canContainData).collect(Collectors.toUnmodifiableList());
```
Comment (jasontedor) [inline server/src/main/java/org/elasticsearch/indices/ShardLimitValidator.java:217]: How about `return node.getRoles().contains(DiscoveryNodeRole.DATA_FROZEN_NODE_ROLE);`? (And sorry if that doesn’t quite compile, I’m reviewing from my iPad, so no IDE support.)

File: server/src/main/java/org/elasticsearch/indices/ShardLimitValidator.java [hunk 3]
```diff
@@ -81,55 +111,119 @@ public void validateShardLimit(final Settings settings, final ClusterState state
...
+
+        return dataRoles.contains(DiscoveryNodeRole.DATA_FROZEN_NODE_ROLE);
+    }
+
+    private static boolean hasNonFrozen(DiscoveryNode node) {
+        final List<DiscoveryNodeRole> dataRoles =
+            node.getRoles().stream().filter(DiscoveryNodeRole::canContainData).collect(Collectors.toUnmodifiableList());
```
Comment (jasontedor) [inline server/src/main/java/org/elasticsearch/indices/ShardLimitValidator.java:224]: `return node.getRoles().stream().anyMatch(r -> r.canContainData() && r !=DiscoveryNodeRole.DATA_FROZEN_NODE_ROLE);`?

Comment (jasontedor) [reply server/src/main/java/org/elasticsearch/indices/ShardLimitValidator.java:224]: This suggested predicate almost exactly matches the name of the method (it has a non-frozen data role), while the current implementation requires a little bit of though. And we can avoid the unnecessary materialization of a collection, etc.

**Review by jasontedor — APPROVED** [review]
This is so well executed, all I have are small nits. Thanks for making this one so easy to review. My small nits don’t require another round.

File: server/src/main/java/org/elasticsearch/indices/ShardLimitValidator.java (hunk 2 above)
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/indices/ShardLimitValidator.java]: Thanks, much nicer and did compile, changed in 67e92fdfad17a875639dcdc3b2dd41f98b9bf8fe

File: server/src/main/java/org/elasticsearch/indices/ShardLimitValidator.java (hunk 3 above)
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/indices/ShardLimitValidator.java]: Thanks, that is nicer, added in 67e92fdfad17a875639dcdc3b2dd41f98b9bf8fe

**Comment by henningandersen** [comment]
@elasticmachine update branch

**Comment by henningandersen** [comment]
@elasticmachine update branch

**Pull request merged**